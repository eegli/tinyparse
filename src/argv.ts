import { Parser } from './parser';
import { PositionalArgs, SimpleRecord, WithPositionalArgs } from './types';
import { parseJSONFile } from './utils';

type TransformArgs = {
  aliases: Map<string, string>;
  filePathFlag?: string;
};

export class ArgvParser<T extends SimpleRecord> extends Parser<T> {
  public build(input: T, positionals: string[]): WithPositionalArgs<T> {
    return {
      ...input,
      _: positionals,
    };
  }
  public transform(
    argv: string[],
    { aliases, filePathFlag }: TransformArgs
  ): [Partial<T>, PositionalArgs] {
    const flagMap = new Map<string, unknown>();

    const positionals: string[] = [];
    let isPositional = true;

    for (let i = 0; i < argv.length; i++) {
      let curr = argv[i];

      if (!curr.startsWith('-') && isPositional) {
        positionals.push(curr);
        continue;
      }

      isPositional = false;

      // Convert short flag to long flag
      if (aliases.has(curr)) {
        curr = '--' + aliases.get(curr);
      }

      // Ignore non-flags
      if (!curr.startsWith('--')) {
        continue;
      }

      let flag = curr.slice(2);
      if (aliases.has(flag)) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        flag = aliases.get(flag)!;
      }

      const flagVal = argv[i + 1];

      // Parse a file
      if (filePathFlag?.slice(2) === flag) {
        parseJSONFile(flagVal).forEach(([key, content]) =>
          flagMap.set(key, content)
        );
      }
      // Assume boolean flag
      else if (!flagVal || flagVal.startsWith('-')) {
        flagMap.set(flag, true);
        // Assume number
        // Assume string
      } else {
        flagMap.set(flag, flagVal);
      }
    }

    return [Object.fromEntries(flagMap) as Partial<T>, positionals];
  }
}
