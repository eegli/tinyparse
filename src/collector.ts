import { WithPositionalArgs } from './types';

export class Collector<T> {
  constructor(private readonly defaults: T) {}
  public collectWithPositionals(positionals: string[]): WithPositionalArgs<T> {
    return {
      ...this.collect(),
      _: positionals,
    };
  }

  public collect(): T {
    return this.defaults;
  }
}
