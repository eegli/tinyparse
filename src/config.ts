import {
  AnyGlobal,
  CommandOptionsMap,
  DefaultHandler,
  ErrorHandler,
  FlagOptionsMap,
  FlagValueRecord,
  MetaOptions,
  SubparserOptionsMap,
} from './types';

interface CoreConfig<O extends FlagValueRecord, G extends AnyGlobal> {
  meta: MetaOptions;
  options: FlagOptionsMap;
  commands: CommandOptionsMap<O, G>;
  parsers: SubparserOptionsMap;
}

export interface CommonConfig<O extends FlagValueRecord, G extends AnyGlobal>
  extends CoreConfig<O, G> {
  defaultHandler: DefaultHandler<O, G>;
  globalSetter?: (options: O) => G | Promise<G>;
  errorHandler?: ErrorHandler;
}

export interface HelpPrinterConfig<
  O extends FlagValueRecord,
  G extends AnyGlobal,
> extends Partial<CoreConfig<O, G>> {}
