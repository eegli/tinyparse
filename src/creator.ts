export type ConfigLike = Record<string, string | number | boolean>;

export function argvToObj(args: string[]): ConfigLike {
  return args.reduce((acc, curr, idx, orig) => {
    if (curr.startsWith('--')) {
      const arg = curr.slice(2);
      let argVal: string | number | boolean = orig[idx + 1];
      // Assume boolean flag
      if (!argVal || argVal.startsWith('--')) {
        acc[arg] = true;
        // Assume number
      } else if (/^\d+$/.test(argVal)) {
        acc[arg] = +argVal;
        // Assume string
      } else {
        acc[arg] = argVal;
      }
    }
    return acc;
  }, {} as ConfigLike);
}

export function configFactory<T extends ConfigLike, R extends keyof T = string>(
  base: T,
  ...required: Array<R extends string ? R : never>
) {
  return function (args?: Partial<T> | Array<string>): Promise<T> {
    return new Promise((resolve, reject) => {
      if (!args) {
        return resolve(base);
      }
      const cfmap = new Map(Object.entries(base));

      // [--id, 123, --secret, xyz] --> arg, argVal, arg, argVal

      if (!Array.isArray(args)) {
        Object.entries(args).forEach(([arg, argVal]) => {
          if (cfmap.has(arg)) {
            const configVal = cfmap.get(arg);
            // Convert string to number if valid
            if (typeof configVal === 'number' && /^\d+$/.test(argVal)) {
              cfmap.set(arg, +argVal);
            } else if (typeof configVal === typeof argVal) {
              cfmap.set(arg, argVal);
            } else {
              reject(
                `Invalid type for option "${arg}". Expected ${typeof configVal}, got ${typeof argVal}`
              );
            }
          } else {
            console.warn(`Ignoring unknown option "${arg}"`);
          }
        });
      } else {
        args.forEach((val, idx, orig) => {
          if (val.startsWith('--')) {
            const arg = val.slice(2);
            const key = cfmap.get(arg);
            if (cfmap.has(arg)) {
              let argVal: string | number | boolean = orig[idx + 1];
              // Boolean flag that default to false
              if (typeof key === 'boolean') {
                cfmap.set(arg, true);
              }
              // Convert string input to number
              else if (typeof key === 'number' && /^\d+$/.test(argVal)) {
                cfmap.set(arg, +argVal);
              } else if (typeof key === typeof argVal) {
                cfmap.set(arg, argVal);
              } else {
                reject(
                  `Invalid type for option "--${arg}". Expected ${typeof key}, got ${typeof argVal}`
                );
              }
            } else {
              console.warn(`Ignoring unknown option "--${arg}"`);
            }
          }
        });
      }

      // Check for required properties
      required.forEach((prop) => {
        if (!cfmap.get(prop)) {
          reject(`Missing required config property "${prop}"`);
        }
      });

      return resolve(Object.fromEntries(cfmap) as T);
    });
  };
}
