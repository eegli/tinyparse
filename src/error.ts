export class ValidationError extends Error {
  constructor(err: string) {
    super(err);
    this.name = 'ValidationError';
    Object.setPrototypeOf(this, ValidationError.prototype);
  }
}
