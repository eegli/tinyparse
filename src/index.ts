import { displayHelp } from './help';
import { parseObjectLiteral } from './parse';
import { transformArgv, transformOptions } from './transform';
import {
  ParsingOptions,
  PositionalArgs,
  SimpleRecord,
  WithPositionalArgs,
} from './types';

export { ValidationError } from './error';

/**
 * Parser factory function. Returns a parser that is bound to the
 * given default values and options per key.
 *
 * @export
 */
export function createParser<T extends SimpleRecord>(
  defaultValues: T,
  parsingOptions?: ParsingOptions<T>
) {
  const options = transformOptions(parsingOptions);
  const filePathArg = parsingOptions?.filePathArg;

  function parse(input?: Partial<T>): Promise<T>;
  function parse(input?: string[]): Promise<WithPositionalArgs<T>>;
  function parse(input: Partial<T> | string[] = {}): Promise<T> {
    let positionalArgs: PositionalArgs | undefined;
    if (Array.isArray(input)) {
      [input, positionalArgs] = transformArgv<T>({
        argv: input,
        filePathFlag: filePathArg?.longFlag,
        options,
      });
    }
    return parseObjectLiteral({
      defaultValues,
      input,
      options,
      forwardArgs: positionalArgs && { _: positionalArgs },
    });
  }

  return {
    help: function (title?: string): string {
      return displayHelp({ defaultValues, options, title, filePathArg });
    },
    parse,
  };
}
