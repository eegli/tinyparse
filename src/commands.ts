import { CommonConfig } from './config';
import { Parser } from './parser';
import {
  AnyGlobal,
  CommandArgPattern,
  DefaultHandler,
  DowncastFlag,
  ErrorHandler,
  FlagOptions,
  FlagOptionsExt,
  FlagValue,
  FlagValueRecord,
  MetaOptions,
  Subcommand,
  Subparser,
} from './types/internals';

export class CommandBuilder<Options, Globals> {
  #config: CommonConfig = {
    meta: {},
    options: new Map(),
    commands: new Map(),
    parsers: new Map(),
    defaultHandler: () => {},
  };

  #takenFlags = new Set<string>();
  #takenCommands = new Set<string>();

  #validateCommand = (command: string): void => {
    if (this.#takenCommands.has(command)) {
      throw new Error(`Command "${command}" has been declared twice`);
    }
  };

  #validateFlag = (flag: string): void => {
    if (this.#takenFlags.has(flag)) {
      throw new Error(`Flag "${flag}" has been declared twice`);
    }
  };

  #validateOption = (key: string): void => {
    if (this.#config.options.has(key)) {
      throw new Error(`Option "${key}" has been declared twice`);
    }
  };

  #validateOneOf = <T>(id: string, type: T, ...options: T[]): void => {
    const expectedType = typeof type;
    const wrongType = options.find((opt) => typeof opt !== expectedType);
    if (wrongType) {
      throw new Error(
        `OneOf for option "${id}" contains invalid type ${typeof wrongType}, expected ${expectedType}`,
      );
    }
  };

  #tryRegisterCommandToken = (command?: string) => {
    if (command) {
      this.#validateCommand(command);
      this.#takenCommands.add(command);
    }
  };

  #tryRegisterFlagToken = (flag?: string) => {
    if (flag) {
      this.#validateFlag(flag);
      this.#takenFlags.add(flag);
    }
  };

  /**
   * Add an option (flag)
   */
  option<
    K extends string,
    V extends FlagValue,
    R extends boolean,
    O extends unknown[],
    Choices = O[number],
    Values = R extends true ? Choices : V | Choices,
  >(
    key: K,
    opts: FlagOptionsExt<V, R, Choices>,
  ): CommandBuilder<Options & Record<K, Values>, Globals>;
  option<K extends string, V extends FlagValue, R extends boolean>(
    key: K,
    opts: FlagOptions<V, R>,
  ): CommandBuilder<Options & Record<K, DowncastFlag<V>>, Globals>;
  option<K extends string, V extends FlagValue, R extends boolean>(
    key: K,
    opts: FlagOptions<V, R>,
  ) {
    const { longFlag, shortFlag, oneOf, defaultValue } = opts;

    if (oneOf) {
      this.#validateOneOf(key, defaultValue, ...oneOf);
    }

    this.#validateOption(key);
    this.#config.options.set(key, opts);

    this.#tryRegisterFlagToken(longFlag);
    this.#tryRegisterFlagToken(shortFlag);

    return this as CommandBuilder<
      Options & Record<K, DowncastFlag<V>>,
      Globals
    >;
  }

  /**
   * Add a subcommand
   */
  subcommand<A extends CommandArgPattern>(
    command: string,
    opts: Subcommand<Options, Globals, A>,
  ) {
    this.#tryRegisterCommandToken(command);
    this.#config.commands.set(
      command,
      opts as Subcommand<FlagValueRecord, AnyGlobal, A>,
    );
    return this;
  }

  /**
   * Add a subparser
   */
  subparser(command: string, opts: Subparser) {
    this.#tryRegisterCommandToken(command);
    this.#config.parsers.set(command, opts);
    return this;
  }

  /**
   * Set the globals
   */
  setGlobals<G extends AnyGlobal>(
    setGlobals: (options: Options) => G | Promise<G>,
  ) {
    this.#config.globalSetter = setGlobals as (
      options: FlagValueRecord,
    ) => AnyGlobal | Promise<AnyGlobal>;
    return this as CommandBuilder<Options, Globals & G>;
  }

  /**
   * Set metadata
   */
  setMeta(meta: MetaOptions) {
    this.#tryRegisterCommandToken(meta.help?.command);
    this.#tryRegisterCommandToken(meta.version?.command);

    this.#tryRegisterFlagToken(meta.help?.longFlag);
    this.#tryRegisterFlagToken(meta.help?.shortFlag);
    this.#tryRegisterFlagToken(meta.version?.longFlag);
    this.#tryRegisterFlagToken(meta.version?.shortFlag);

    this.#config.meta = meta;
    return this;
  }

  /**
   * Set the error handler
   */
  onError(handler: ErrorHandler) {
    this.#config.errorHandler = handler;
    return this;
  }

  /**
   * Set the default handler
   */
  defaultHandler(handler?: DefaultHandler<Options, Globals>) {
    return new Parser<Options>({
      ...this.#config,
      defaultHandler:
        (handler as DefaultHandler) || this.#config.defaultHandler,
    });
  }
}
