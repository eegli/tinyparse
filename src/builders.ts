import { transformArgv } from './argv';
import { validateCommandArgs } from './commands';
import { collect } from './flags';
import {
  Subcommand,
  Downcast,
  FlagOptionMap,
  FlagRecord,
  FlagOptions,
  FlagArgValue,
  CommandOptionMap,
  CommandArgPattern,
} from './types';

export class CommandBuilder<F extends FlagRecord = Record<never, never>> {
  #flags: FlagOptionMap = new Map();
  #commands: CommandOptionMap<F> = new Map();

  public subcommand<P extends CommandArgPattern>(
    command: string,
    opts: Subcommand<F, P>,
  ) {
    this.#commands.set(command, opts);
    return this;
  }

  public flag<T extends string, V extends FlagArgValue>(
    flag: T,
    opts: FlagOptions<V>,
  ) {
    this.#flags.set(flag, opts);
    // TODO: Figure out how to make this typecheck properly
    return this as unknown as CommandBuilder<F & Record<T, Downcast<V>>>;
  }

  public build() {
    return new DispatchBuilder<F>(this.#flags, this.#commands);
  }
}

export class DispatchBuilder<F extends FlagRecord> {
  #handler?: (flags: F, positionals: string[]) => void;
  #flags: FlagOptionMap;
  #commands: CommandOptionMap<F>;

  constructor(flags: FlagOptionMap, commands: CommandOptionMap<F>) {
    this.#flags = flags;
    this.#commands = commands;
  }

  public default<T extends F>(
    handler: T extends FlagRecord
      ? (flags: T, positionals: string[]) => void
      : never,
  ) {
    this.#handler = handler;
    return this;
  }

  public parse(argv: string[]): {
    flags: F;
    call: () => void;
  } {
    const [flagMap, positionals] = transformArgv(argv);
    const flags = collect(flagMap, this.#flags) as F;

    const [command, ...args] = positionals;
    const commandOpts = this.#commands.get(command);

    if (command && commandOpts) {
      validateCommandArgs(command, commandOpts, args);
      const call = commandOpts.handler.bind(this, flags, args);
      return {
        flags,
        call,
      };
    }
    if (this.#handler) {
      const call = this.#handler?.bind(this, flags, positionals);
      return {
        flags,
        call,
      };
    }
    return {
      flags,
      call: () => {},
    };
  }
}
