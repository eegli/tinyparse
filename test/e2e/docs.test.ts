import type { ParserOptions, Value } from '../../src';
import { createParser, ValidationError } from '../../src';
import { mockFs } from '../_setup';

describe('Docs', () => {
  test('cli arguments, keeps defaults', async () => {
    const { parse } = createParser({ hello: 'world' });
    let parsed = await parse();
    expect(parsed).toStrictEqual({ _: [], hello: 'world' });

    parsed = await parse(['--hello', 'john']);
    expect(parsed).toStrictEqual({ _: [], hello: 'john' });
  });
  test('cli arguments, internal validation', () => {
    expect(() => {
      createParser(
        { a: '', b: '' },
        { options: { a: { shortFlag: 'a' }, b: { shortFlag: 'a' } } },
      );
    }).toThrow(
      'Parser config validation error, conflicting short flag: -a has been declared twice. Check your settings for short flags.',
    );
  });
  test('cli arguments, positional arguments 1', async () => {
    const { parse } = createParser({});
    const parsed = await parse(['hello-world']);
    expect(parsed).toStrictEqual({ _: ['hello-world'] });
  });

  test('cli arguments, positional arguments 2', () => {
    const { parseSync } = createParser(
      {},
      {
        positionals: {
          count: '=0', // Allow no positional arguments
        },
      },
    );
    expect(() => {
      parseSync(['hello-world']);
    }).toThrow('Invalid number of positional arguments: Expected 0, got 1');
  });

  test('cli arguments, positional arguments 3', () => {
    const { parseSync } = createParser(
      {},
      {
        positionals: {
          count: '>=1', // Require 1 or more positional arguments
        },
      },
    );
    expect(() => {
      parseSync([]);
    }).toThrow(
      'Invalid number of positional arguments: Expected at least 1, got 0',
    );
  });

  test('cli arguments, positional arguments 4', () => {
    const { parseSync } = createParser(
      {},
      {
        positionals: {
          count: '<=1', // Require 1 or less positional arguments
        },
      },
    );
    expect(() => {
      parseSync(['a', 'b']);
    }).toThrow(
      'Invalid number of positional arguments: Expected at most 1, got 2',
    );
  });

  test('cli arguments, boolean flags 1', async () => {
    const { parse } = createParser({ verbose: false });
    const parsed = await parse(['--verbose']);
    expect(parsed.verbose).toBe(true);
    expect(parsed).toStrictEqual({ verbose: true, _: [] });
  });
  test('cli arguments, boolean flags 2', async () => {
    const { parse } = createParser({ verbose: true });
    const parsed = await parse(['--verbose']);
    expect(parsed.verbose).toBe(true);
  });
  test('cli arguments, number conversion', async () => {
    const { parse } = createParser({
      followers: -1, // expect number
      year: '2000', // expect (date) string
    });
    const parsed = await parse(['--followers', '8', '--year', '2023']);
    expect(parsed.followers).toBe(8);
    expect(parsed.year).toBe('2023');
  });

  test('custom flags', async () => {
    const { parse } = createParser(
      {
        userName: '',
        verbose: false,
      },
      {
        options: {
          userName: {
            longFlag: 'user',
          },
          verbose: {
            shortFlag: 'v',
          },
        },
      },
    );
    const parsed = await parse(['-v', '--user=john']);
    expect(parsed.verbose).toBe(true);
    expect(parsed.userName).toBe('john');
  });

  test('custom validation', () => {
    const { parseSync } = createParser(
      { birthDate: '2000-01-01' },
      {
        options: {
          birthDate: {
            longFlag: 'bday',
            customValidator: {
              isValid(value): value is Value {
                if (typeof value !== 'string') return false;
                return !isNaN(new Date(value).getTime());
              },
              errorMessage(value, flag) {
                return `Invalid value '${value}' for option '${flag}'. Expected a valid date string`;
              },
            },
          },
        },
      },
    );
    // Valid date string
    expect(() => {
      parseSync(['--bday', '2000-01-01']);
    }).toBeTruthy();

    // What a weird month...
    expect(() => {
      parseSync(['--bday', '2000-22']);
    }).toThrow(
      "Invalid value '2000-22' for option '--bday'. Expected a valid date string",
    );
  });

  test('decamelization', async () => {
    const { parse } = createParser(
      { userName: '' },
      {
        decamelize: true,
      },
    );
    const parsed = await parse(['--user-name', 'eegli']);

    expect(parsed.userName).toBe('eegli');
  });

  test('file reading, valid', () => {
    const file = {
      userName: 'eegli',
      hasGitHubPlus: true,
    };
    mockFs.readFileSync.mockImplementationOnce((path) => {
      if (path === 'github.json') {
        return JSON.stringify(file);
      }
      throw new Error();
    });
    const { parseSync } = createParser(
      {
        userName: '',
        hasGitHubPlus: false,
      },
      {
        filePathArg: {
          longFlag: 'config',
          shortFlag: 'c',
          description: 'Path to your Github config file',
        },
      },
    );

    const parsed = parseSync(['-c', 'github.json']);

    expect(parsed.userName).toBe('eegli');
    expect(parsed.hasGitHubPlus).toBe(true);
  });
  test('file reading, invalid', () => {
    const file = {
      userName: {
        name: 'eegli',
      },
    };
    mockFs.readFileSync.mockImplementationOnce((path) => {
      if (path === 'bad-github.json') {
        return JSON.stringify(file);
      }
      throw new Error();
    });
    const { parseSync } = createParser(
      {
        userName: '',
      },
      {
        filePathArg: {
          longFlag: 'config',
        },
      },
    );

    expect(() => {
      parseSync(['--config', 'bad-github.json']);
    }).toThrow(`Invalid type for userName. "[object Object]" is not a string`);
  });
  test('printing args, without decamelization', () => {
    const { help } = createParser(
      {
        userName: '',
        age: Infinity,
        hasGithubProfile: false,
      },
      {
        options: {
          userName: {
            shortFlag: 'u',
            description: 'Your GitHub username',
          },
          hasGithubProfile: {
            description: 'Indicate whether you have a Github profile',
          },
          age: {
            required: true,
          },
        },
        filePathArg: {
          longFlag: 'config',
          description: 'Path to your Github config file',
        },
      },
    );
    const helpText = help({
      title: 'CLI usage',
      base: 'my-cli <message> [flags]',
    });
    expect(helpText).toMatchInlineSnapshot(`
      "CLI usage

      my-cli <message> [flags]

      Required flags
         --age [number]

      Optional flags
         --hasGithubProfile [boolean]
         Indicate whether you have a Github profile

         -u, --userName [string]
         Your GitHub username

         --config [string]
         Path to your Github config file
      "
    `);
  });

  test('printing args, with decamelization', () => {
    const { help } = createParser(
      {
        userName: '',
        hasGithubProfile: false,
      },
      {
        decamelize: true,
        options: {
          userName: {
            longFlag: 'user',
          },
        },
      },
    );
    const helpText = help();
    expect(helpText).toMatchInlineSnapshot(`
      "Usage

      Optional flags
         --has-github-profile [boolean]

         --user [string]"
    `);
  });

  test('error handling, rejects for missing args', () => {
    const { parseSync } = createParser(
      { username: '' },
      {
        options: {
          username: {
            required: true,
          },
        },
      },
    );

    expect(() => {
      parseSync(); // Whoops, forgot username!
    }).toThrow(new ValidationError('Missing required argument username'));
  });
  test('error handling, rejects invalid types', () => {
    const { parseSync } = createParser({ age: 0 });

    expect(() => {
      parseSync(['--age']);
    }).toThrow(
      new ValidationError('Invalid type for --age. "true" is not a number'),
    );
  });
  // eslint-disable-next-line jest/expect-expect
  test('typescript, bootstrapping', () => {
    const defaults = {
      abc: 'abc',
    };

    type CustomOptions = ParserOptions<typeof defaults>;

    // Construct the options for a parser...
    const options: CustomOptions = {
      options: {
        abc: {
          /* ... */
        },
      },
    };
    // ...and bootstrap it later
    createParser(defaults, options);
  });
});
