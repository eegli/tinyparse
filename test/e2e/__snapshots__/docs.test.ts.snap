// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error handling catches error 1`] = `
[
  "Missing required option --foo",
  "Usage: my-app [command] <...flags>

Required flags
   --foo [boolean]

To view this help message, run "my-app help"",
]
`;

exports[`help calls help printer 1`] = `
"A brief description of my-cli

Usage: my-cli [command] <...flags>

Commands
   baz <arg>
   - Baz command

Required flags
   -f, --foo [string]
   Foo option

Optional flags
   --bar [date]
   Foo option

To view this help message, run "my-cli help" or add --help or -h to any command"
`;
